test/test_util.js

```diff
@@ -1,213 +1,109 @@
+  const input = view.getByRole('textbox');
 
-  event = { ...event, ...eventProps, target };
+  const user = userEvent.setup();
 
-  return event;
+  return { ...view, view: view, input, user };
 }
 
-export function setCaretPosition(event, caretPos) {
-  const { target } = event;
-  target.focus();
-  target.setSelectionRange(caretPos, caretPos);
-}
+export async function simulateKeyInput(
+  user,
+  input,
+  key,
+  selectionStart = 0,
+  selectionEnd,
+  options,
+) {
+  const v = input.value;
+
+  let [start, end] = [selectionStart, selectionEnd ?? selectionStart];
+  if (selectionStart > v.length) {
+    start = v.length;
+  }
+  if (end > v.length) {
+    end = v.length;
+  }
 
-export function simulateKeyInput(input, key, selectionStart, selectionEnd, setSelectionRange) {
-  if (selectionEnd === undefined) {
-    selectionEnd = selectionStart;
+  if (key.length === 0) {
+    return;
   }
 
-  const currentValue = input.prop('value');
-  let defaultPrevented = false;
-
-  const keydownEvent = getEvent(
-    {
-      preventDefault: function () {
-        defaultPrevented = true;
-      },
-      key,
-      isUnitTestRun: true,
-      persist: persist.bind(null, 'keydown'),
-    },
-    {
-      value: currentValue,
-      selectionStart,
-      selectionEnd,
-      setSelectionRange: setSelectionRange || noop,
-      focus: noop,
-    },
-  );
-
-  //fire key down event
-  input.simulate('keydown', keydownEvent);
-
-  //fire change event
-  if (!defaultPrevented && key !== 'ArrowLeft' && key !== 'ArrowRight') {
-    //get changed caret positon
-    let newCaretPosition, newValue;
-
-    if (key === 'Backspace') {
-      newCaretPosition = selectionStart !== selectionEnd ? selectionStart : selectionStart - 1;
-      newValue =
-        selectionStart !== selectionEnd
-          ? currentValue.substring(0, selectionStart) +
-            currentValue.substring(selectionEnd, currentValue.length)
-          : currentValue.substring(0, newCaretPosition) +
-            currentValue.substring(selectionStart, currentValue.length);
-    } else if (key === 'Delete') {
-      newCaretPosition = selectionStart;
-      newValue =
-        selectionStart !== selectionEnd
-          ? currentValue.substring(0, selectionStart) +
-            currentValue.substring(selectionEnd, currentValue.length)
-          : currentValue.substring(0, selectionStart) +
-            currentValue.substring(selectionStart + 1, currentValue.length);
-    } else {
-      newCaretPosition = selectionStart + key.length;
-      newValue =
-        selectionStart !== selectionEnd
-          ? currentValue.substring(0, selectionStart) +
-            key +
-            currentValue.substring(selectionEnd, currentValue.length)
-          : currentValue.substring(0, selectionStart) +
-            key +
-            currentValue.substring(selectionStart, currentValue.length);
-    }
+  const specialKeys = ['{Backspace}', '{Delete}', '{ArrowLeft}', '{ArrowRight}'];
```

Comment:
Q: Why do we need manual handling for specialKeys? Any issue with testing library? If yes then can we add it on comment.

Reply:
It's not the special keys but all other keys that need conditional handling (the next few lines on code should probably make this clear).

I have not faced this issue with testing-library elsewhere but cannot say with certainty what the root cause is. Consider this a workaround for now. When time permits, I will try and come back to this and dig deeper.

I will add a brief comment to explain this.

Reply:
Yes, most likely its issue with jsdom + testing lib.